test/test_integration.py

```diff
@@ -0,0 +1,108 @@
+        cwl_local_path = os.path.abspath('testdata/md5sum.cwl')
+        output_filepath = run_md5sum(cwl_input='file://' + cwl_local_path)
+
+        self.assertTrue(check_for_file(output_filepath), 'Output file was not found: ' + str(output_filepath))
+        shutil.rmtree('workflows')
+
+
+def run_md5sum(cwl_input):
+    """Pass a local md5sum cwl to the wes-service server, and return the path of the output file that was created."""
+    endpoint = 'http://localhost:8080/ga4gh/wes/v1/workflows'
+    params = {'output_file': {'path': '/tmp/md5sum.txt', 'class': 'File'}, 'input_file': {'path': '../../testdata/md5sum.input', 'class': 'File'}}
+    body = {'workflow_url': cwl_input, 'workflow_params': params, 'workflow_type': 'CWL', 'workflow_type_version': 'v1.0'}
+    response = requests.post(endpoint, json=body).json()
+    output_dir = os.path.abspath(os.path.join('workflows', response['workflow_id'], 'outdir'))
+    return os.path.join(output_dir, 'md5sum.txt')
+
+
+def get_server_pids():
+    try:
+        pids = subprocess.check_output(['pgrep', '-f', 'wes_service_main.py']).strip().split()
+    except subprocess.CalledProcessError:
+        return None
+    return pids
+
+
+def check_for_file(filepath, seconds=20):
+    """Return True if a file exists within a certain amount of time."""
+    if os.path.exists(filepath):
+        return True
+    else:
+        wait_counter = 0
+        while not os.path.exists(filepath):
+            time.sleep(1)
+            wait_counter += 1
+            if os.path.exists(filepath):
+                return True
+            if wait_counter > seconds:
+                return False
+
+class CwltoolTest(IntegrationTest):
+    """Test using cwltool."""
+    def setUp(self):
+        """
+        Start a (local) wes-service server to make requests against.
+        Use cwltool as the wes-service server 'backend'.
+        """
+        self.wes_server_process = subprocess.Popen('python {}'.format(os.path.abspath('wes_service/wes_service_main.py')),
+                                                   shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+        time.sleep(5)
+
+class ToilTest(IntegrationTest):
+    """Test using Toil."""
+    def setUp(self):
+        """
+        Start a (local) wes-service server to make requests against.
+        Use toil as the wes-service server 'backend'.
+        """
+        self.wes_server_process = subprocess.Popen('python {} '
+                                                   '--opt runner=cwltoil --opt extra=--logLevel=CRITICAL'
+                                                   ''.format(os.path.abspath('wes_service/wes_service_main.py')),
+                                                   shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+        time.sleep(5)
+
+# Prevent pytest/unittest's discovery from attempting to discover the base test class.
+del IntegrationTest
```

Comment:
Can you explain what's going on here?

Reply:
This class inherits IntegrationTest and overwrites the setup method with one that spins up a Toil wes-service. The same for cwltools.

I wanted to run the same tests with toil and cwltool. So I created a class they could each inherit and left a setup in each to overwrite the original and spin up either cwltool or toil respectively.

The del at the bottom prevents unittest from "seeing" the inherited class and so it doesn't run. Only the intended classes.

Reply:
It's probably better to specify the abstract class in another file that isn't titled `test_`. Then you can use and reuse the class without running it with the test suite.

If you want to skip tests in pytest, a decorator is used: https://docs.pytest.org/en/latest/skipping.html

Reply:
I follow the decorator convention, though I've only used it to skip tests based on requirements (skip if no docker installation, or if no internet, etc.).  I've never seen it applied to an abstract base class.  If you have a good example of this, I'd be inclined to use it.

For me, intuitively at least, I like keeping the ABC in the same file since it makes it easier to read and see what's going on.  I'd like to argue for it, since, in my opinion, it's more compact and more readable.